p2p/peer_behaviour.go

```diff
@@ -0,0 +1,140 @@
+package p2p
+
+import (
+	"errors"
+	"sync"
+)
+
+// ErrorPeerBehaviour are types of observable erroneous peer behaviours.
+type ErrorPeerBehaviour int
+
+const (
+	ErrorPeerBehaviourUnknown = iota
+	ErrorPeerBehaviourBadMessage
+	ErrorPeerBehaviourMessageOutofOrder
+)
+
+// GoodPeerBehaviour are types of observable positive peer behaviours.
+type GoodPeerBehaviour int
+
+const (
+	GoodPeerBehaviourVote = iota + 100
+	GoodPeerBehaviourBlockPart
```

Comment:
This kind of knowledge shouldn't be in the peer package - ie. the p2p package shouldn't depend on (or have knowledge of) consensus reactor behaviour.

Do we need per-channel peer behaviour? Then we can let these behaviours be defined where the channels are set up (eg. in the specific reactors)

(Recall each reactor can utilize multiple channels, where each channel is a stream multiplexed on the underlying MConn)

Reply:
We can't really express this with the go type system and it might likely lead to hard to debug code paths again, we wouldn't gain much, compared to the current implementation.

One approach is to extrat all possible behvaviours into it's own package.

Reply:
> One approach is to extrat all possible behvaviours into it's own package.

This actually might not be a bad idea. A higher level package that deals with all these things. My main concern is just to avoid having reactor-specific ideas trickle into the `p2p` package.

Reply:
I think this makes sense. This could be moved to a `behaviour` package which effectively maps the reactor specific events to peer management.

I wonder if we couldn't implement a similar pattern as consensus message handling in [adr-30]("https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-030-consensus-refactor.md") 🤔 

Implementation behaviour/peer_behaviour.go:
```go
struct PeerBehaviour interface {}
...

type PeerBehaviourReporter interface {
	Report(peerID p2p.ID, behaviour PeerBehaviour) error
}
...
func (spbr *SwitchPeerBehaviourReporter) Report(peerID ID, behaviour PeerBehaviour) error {
	...
	switch behaviour.(type) {
	case PeerBehaviourVote
		spbr.sw.MarkPeerAsGood(peer)
		break
	...
	}
}
```


Example consensus/reactor.go:
```go
type PeerBehaviourVote struct {...}
...
peerReporter.Report(peerID, PeerBehaviourVote{...})
```

This would also leave room for reactors to provide richer information to the reporter in reactor specific ways. Thoughts?

Reply:
I think this makes sense. This could be moved to a `behaviour` package which effectively maps the reactor specific events to peer management.

I wonder if we couldn't implement a similar pattern as the consensus message handling in [adr-30](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-030-consensus-refactor.md) 🤔 

Implementation behaviour/peer_behaviour.go:
```go
struct PeerBehaviour interface {}
...

type PeerBehaviourReporter interface {
	Report(peerID p2p.ID, behaviour PeerBehaviour) error
}
...
func (spbr *SwitchPeerBehaviourReporter) Report(peerID ID, behaviour PeerBehaviour) error {
	...
	switch behaviour.(type) {
	case PeerBehaviourVote
		spbr.sw.MarkPeerAsGood(peer)
		break
	...
	}
}
```


Example consensus/reactor.go:
```go
type PeerBehaviourVote struct {...}
...
peerReporter.Report(peerID, PeerBehaviourVote{...})
```
This might also resolve the [ #3552 (comment)](https://github.com/tendermint/tendermint/pull/3552#discussion_r280940857) and allow  reactor specific behaviour. Thoughts?